### DISCUSSION TITLE
Framework Approach

### DISCUSSION THREAD

**Discussion by WendellAdriel** [minicli/minicli#81] [Ideas]
## Introduction

Minicli core idea is to be a minimal package to create CLI apps without any dependencies and that's wonderful. A lot can be accomplished with this, but sometimes we need a more robust solution due to the complexity that some apps require.

## Proposal

We could create a new repository: `minicli/framework` that will be built on top of the `minicli/minicli` core to extend some features like having a plugin system proposed on https://github.com/orgs/minicli/discussions/80, giving the ability for users to provide their own DI container and so much more.

The core idea of being a minimal framework would maintain and we would add only dependencies that are REALLY required.

**Comment by erikaheidi** [comment]
I love this idea. As you mentioned, being dependency-free is a core value of `minicli/minicli`, so we need to keep it that way - which doesn't prevent us from creating a few optional layers of abstraction that will give users a better development experience when dealing with bigger applications. This way, the core package can still be used for simple things and one-file apps without carrying other dependencies.

minicli/framework sounds legit to me 👍🏼

**Reply by JustSteveKing** [reply]
Consider it done: https://github.com/minicli/framework

**Reply by JustSteveKing** [reply]
I'm going to take some inspiration from Cobra (a go package) for this: https://github.com/spf13/cobra

The main one being where you register a root command, and then add additional commands to it.

**Comment by WendellAdriel** [comment]
Since we already have the repository at https://github.com/minicli/framework
I'm going to close this one and then we can take on the framework repository issues to work on it
